Initial draft of article on how to use graphical tiles (#2)
Thraka/thraka.github.io: articles/how-to-use-graphical-tiles.md

> # How to use Graphical Tiles

Reviewer: ```suggestion
---
description: Learn how to add a graphical font tileset to SadConsole.
ms.date: 10/10/2019
---

# How to use Graphical Tiles in SadConsole
```

Next, create an empty spritesheet in your graphical editing software of choice. Ideally, include a grid, so you can easily distinguish tiles. Draw or copy/paste a few characters into the appropriate position. If you're not sure which cell corresponds to which character, you can refer to the existing font files, such as the `IBM_8x16.png` file.

Hide the grid and export the file into your project, under the `Fonts` directory. Next, create the `.font` metadata file. Here's a sample:

```
{
    "FilePath": "MyCustomFont_16x16.png",
    "GlyphHeight": 16,
    "GlyphPadding": 0,
    "GlyphWidth": 16,
    "Name": "MyCustomFont",
    "SolidGlyphIndex": 219
}
```

Replace the font sizes and the filename with appropriate values. Note the `SolidGlyphIndex` value: it's used whenever you print an empty tile, or call a method to fill a console.

## Using the Spritesheet Font In-Game

Somewhere in your game console constructor, set the font:

```
var fontMaster = SadConsole.Global.LoadFont("Fonts/MyCustomFont.font");
var normalSizedFont = fontMaster.GetFont(SadConsole.Font.FontSizes.One);
this.Font = normalSizedFont;

> ```

Reviewer: This can be improved. If you keep loading the font over and over you end up with different texture objects for console's that have already loaded a font. For example.. If you have this in your console constructor and use something like

```csharp
var console1 = new MyConsole();
var console2 = new MyConsole();
```

Each console (in the constructor) is loading a new copy of the font. This means that two textures have been loaded into the graphics card. You should instead load it one time at the start of your game, and then reference it from then on.

```csharp
private static void Init()
{
    SadConsole.Global.LoadFont("Fonts/MyCustomFont.font");

    var console = new SadConsole.Console(10, 10, SadConsole.Global.Fonts["MyCustomFont"].GetFont(Font.FontSizes.One));
}
```

This loads the font into the system and caches it in the `Global.Fonts` collection. The `Name` setting from the font definition file is the id used to reference the font.

If you want your own console type to **always** use the custom font, you should modify the constructor to reference it, since it was loaded at startup

```csharp
private class MyConsole : Console
{
    public MyConsole(int width, int height) : base(width, height, SadConsole.Global.Fonts["MyCustomFont"].GetFont(Font.FontSizes.One))
    {

    }
}
```

Author: This is interesting. My game is one giant console and I never got sub-consoles to work properly (had to end up coding my own work-arounds, which have their own bugs).

I'll update my article to mention doing this in the game constructor and just using the global font by reference. I think that's the most elegant solution.

You're done! Any calls to `SetGlyph` will use the image from your custom font spritesheet. You can also change the enumeration value of `SadConsole.Font.FontSizes` to a different value (eg. `Two` or `Quarter`) to zoom your game in/out or to render at a different size than the sprite image size.

Since SadConsole limits you to one font per console, you need to create sub-consoles for anything you want to render normally, such as a status bar, sub-menu, etc.

## Colouring Sprites

> To colour your sprites, simply specify a `Color` in your `SetGlyph` calls. By default, your font image uses white, so your images will render white. If you specify a color to `SetGlyph`, SadConsole will multiply that colour by the spritesheet colour when it draws.

Reviewer: ```suggestion
To color your sprites, simply specify a `Color` in your `SetGlyph` calls. By default, your font image uses white, so your images will render white. If you specify a color to `SetGlyph`, SadConsole will multiply that color by the spritesheet color when it draws.
```

You can use this for effects; just be aware that drawing pixels in your spritesheet in a color other than white, will restrict what colors they can appear with in-game.

> If you prefer full-color sprites instead, simply colour them as desired within the spritesheet, and always render them as white when you call `SetGlyph`. They will appear exactly as they do in the spritesheet, colors and all.

Reviewer: ```suggestion
If you prefer full-color sprites instead, simply recolor them as desired within the spritesheet, and when you use the glyph in SadConsole, always use the color White for the foreground. They will appear exactly as they do in the spritesheet, colors and all.
```
